File accesscontrol blocks deletion of allowed files

<!--- Please keep this note for other contributors -->

### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

---

### Steps to reproduce
1. Upload allowed file
2. Try to delete the same file

All the stuff below refers to a csv i uploaded but you could also just create a new text file in nextcloud (.md) and try to delete that with the below ruleset.

### Expected behaviour
Usually - and it did so for ages - the access control applies to all actions the same. Meaning if i'm able to create/upload/edit/download/rename i should be able to delete.

### Actual behaviour
Right now i can't delete ANY file. It doesn't matter if it already exists, i create it or i upload a new one.
As soon as i adjust the ruleset so it doesn't match my user i can delete everything again.

Fun fact: folders CAN be deleted - even with files inside them.

Thats the ruleset:
```
When File is accessed
File MIME Type is not Folder
and Group membership is not member of SomeGroup
and Group membership is not member of SomeOtherGroup
and File MIME Type does not match custom mimetype: /^image\/(jpeg|png|gif|bmp|webp|tiff)(?:;.*)?$/i
and File MIME Type does not match custom mimetype: /^text\/(plain|csv|csv\+tsv|html|markdown)(?:;.*)?$/i
and File MIME Type does not match custom mimetype: /^video\/(mp4|webm|ogg|quicktime|mpeg)(?:;.*)?$/i
and File MIME Type does not match custom mimetype: /^audio\/(mpeg|mp3|wav|ogg|flac|x-m4a)(?:;.*)?$/i
and File MIME Type does not match custom mimetype: /^application\/vnd\.apple\.(pages|numbers|keynote)(?:;.*)?$/i
and File MIME Type does not match custom mimetype: /^application\/vnd\.openxmlformats-officedocument\.(wordprocessingml\.document|spreadsheetml\.sheet|presentationml\.(presentation|slideshow))(?:;.*)?$/i
and File MIME Type does not match custom mimetype: /^application\/pdf(?:;.*)?$/i
and File MIME Type does not match custom mimetype: /^application\/(zip|x-7z-compressed|x-gzip|gzip)(?:;.*)?$/i
and File MIME Type does not match custom mimetype: /^application\/(json|xml|yaml)(?:;.*)?$/i
and File MIME Type does not match custom mimetype: /^(application|inode)\/x-empty(?:;.*)?$/i
Block access to a file
```

This is the flow.log entry:
```
{"reqId":"WHRPXaqk42MHN3cZcsi3","level":1,"time":"2025-09-30T10:04:04+02:00","remoteAddr":"X.X.X.X","user":"xxx","app":"workflowengine","method":"DELETE","url":"/remote.php/dav/files/xxx/report-635.csv","message":"Last qualified flow configuration is going to run","userAgent":"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/140.0.0.0 Safari/537.36","version":"31.0.8.1","data":{"operation":{"class":"OCA\\FilesAccessControl\\Operation","name":"Zugriff auf Datei verhindern"},"configuration":{"id":7,"class":"OCA\\FilesAccessControl\\Operation","name":"","checks":"[2,59,49,82,83,84,85,86,87,88,89,90,91]","operation":"deny","entity":"OCA\\WorkflowEngine\\Entity\\File","events":"[]","scope_type":0,"scope_actor_id":""}}}
```
This is the mime type according to different components:
```
apache2 log: text/csv
file -i: text/plain; charset=utf-8
finfo: text/plain
```

So the mimetype should match and it did so since we're using nextcloud. I do test a lot of stuff each time i update the system so i can confirm it 100% worked on 15th of August when i upgraded to 31.0.8. Today i tried upgrading to 31.0.9, did my usual tests and thought the update itself broke things. So i reverted to 31.0.8 (via snapshot) but the problem persists and i was able to narrow it down to the Flow engine.

### Server configuration

**Web server:** Apache2

**Database:** Maria

**PHP version:** 8.3

**Nextcloud version:** 31.0.8 (also happens with 31.0.9)

<details>
<summary>List of activated apps</summary>

```
Enabled:
  - activity: 4.0.0
  - admin_audit: 1.21.0
  - bruteforcesettings: 4.0.0
  - cloud_federation_api: 1.14.0
  - comments: 1.21.0
  - dav: 1.33.0
  - federatedfilesharing: 1.21.0
  - files: 2.3.1
  - files_accesscontrol: 2.0.1
  - files_antivirus: 6.0.5
  - files_automatedtagging: 2.0.0
  - files_downloadlimit: 4.0.0
  - files_pdfviewer: 4.0.0
  - files_reminders: 1.4.0
  - files_retention: 2.0.1
  - files_sharing: 1.23.1
  - files_trashbin: 1.21.0
  - files_versions: 1.24.0
  - impersonate: 2.0.0
  - logreader: 4.0.0
  - lookup_server_connector: 1.19.0
  - notifications: 4.0.0
  - oauth2: 1.19.1
  - ownershiptransfer: 1.3.0
  - password_policy: 3.0.0
  - profile: 1.0.0
  - provisioning_api: 1.21.0
  - quota_warning: 1.22.0
  - registration: 2.8.0
  - related_resources: 2.0.0
  - serverinfo: 3.0.0
  - settings: 1.14.0
  - support: 3.0.0
  - systemtags: 1.21.1
  - terms_of_service: 4.6.0
  - text: 5.0.0
  - theming: 2.6.1
  - theming_customcss: 1.18.0
  - twofactor_admin: 4.8.0
  - twofactor_backupcodes: 1.20.0
  - twofactor_totp: 13.0.0-dev.0
  - updatenotification: 1.21.0
  - user_retention: 1.15.0
  - viewer: 4.0.0
  - workflowengine: 2.13.0
Disabled:
  - app_api: 5.0.2 (installed 4.0.0)
  - circles: 31.0.0 (installed 30.0.0)
  - contactsinteraction: 1.12.0 (installed 1.11.0)
  - dashboard: 7.11.0 (installed 7.10.0)
  - encryption: 2.19.0
  - federation: 1.21.0 (installed 1.20.0)
  - files_external: 1.23.0
  - firstrunwizard: 4.0.0 (installed 3.0.0)
  - nextcloud_announcements: 3.0.0 (installed 2.0.0)
  - photos: 4.0.0 (installed 3.0.2)
  - privacy: 3.0.0 (installed 2.0.0)
  - recommendations: 4.0.0 (installed 3.0.0)
  - sharebymail: 1.21.0 (installed 1.20.0)
  - survey_client: 3.0.0 (installed 2.0.0)
  - suspicious_login: 9.0.1 (installed 8.0.0)
  - twofactor_nextcloud_notification: 5.0.0
  - user_ldap: 1.22.0
  - user_status: 1.11.0 (installed 1.10.0)
  - weather_status: 1.11.0 (installed 1.10.0)
  - webhook_listeners: 1.2.0 (installed 1.1.0-dev)
```
</details>

<details>
<summary>Nextcloud configuration</summary>

```
{
    "system": {
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "***REMOVED SENSITIVE VALUE***",
            "***REMOVED SENSITIVE VALUE***"
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "mysql",
        "version": "31.0.8.1",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "mysql.utf8mb4": true,
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "overwriteprotocol": "https",
        "overwrite.cli.url": "https:\/\/***REMOVED SENSITIVE VALUE***",
        "htaccess.RewriteBase": "\/",
        "memcache.local": "\\OC\\Memcache\\APCu",
        "memcache.distributed": "\\OC\\Memcache\\Redis",
        "memcache.locking": "\\OC\\Memcache\\Redis",
        "redis": {
            "host": "***REMOVED SENSITIVE VALUE***",
            "port": 0,
            "timeout": 0
        },
        "maintenance_window_start": 23,
        "maintenance": false,
        "auth.webauthn.enabled": false,
        "default_phone_region": "DE",
        "reduce_to_languages": [
            "de",
            "en"
        ],
        "default_language": "de",
        "default_locale": "de",
        "default_timezone": "Europe\/Berlin",
        "profile.enabled": false,
        "knowledgebaseenabled": false,
        "session_lifetime": 7200,
        "remember_login_cookie_lifetime": 0,
        "skeletondirectory": "\/opt\/nextcloud\/data\/default",
        "templatedirectory": "",
        "enable_previews": false,
        "loglevel": 2,
        "log.condition": {
            "apps": [
                "admin_audit"
            ]
        },
        "logfile": "\/var\/log\/nextcloud\/nextcloud.log",
        "logfile_audit": "\/var\/log\/nextcloud\/audit.log",
        "logtimezone": "Europe\/Berlin",
        "log_rotate_size": 0,
        "trashbin_retention_obligation": "auto, 7",
        "simpleSignUpLink.shown": false,
        "allowed_admin_ranges": [
            "***REMOVED SENSITIVE VALUE***",
            "***REMOVED SENSITIVE VALUE***"
        ],
        "connectivity_check_domains": [
            "nextcloud.com",
            "***REMOVED SENSITIVE VALUE***"
        ],
        "twofactor_enforced": "true",
        "twofactor_enforced_groups": [
            "user"
        ],
        "twofactor_enforced_excluded_groups": [
            "admin"
        ],
        "mail_smtpmode": "smtp",
        "mail_sendmailmode": "smtp",
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpport": "25",
        "mail_smtpdebug": false,
        "trusted_proxies": "***REMOVED SENSITIVE VALUE***",
        "forbidden_filename_characters": [
            "\\",
            "\/"
        ],
        "forbidden_filename_extensions": [
            ".filepart",
            ".part"
        ]
    }
}
```
</details>

### Browser

**Browser name:** Firefox/Chrome/Edge

**Browser version:**  newest of each

**Operating system:** Windows 11

<details>
<summary>Browser log</summary>

```
index.mjs:51 [WARN] viewer: Some mimes were ignored because they are not enabled in the server previews config Object
log @ index.mjs:51
deleteUtils.ts:99  DELETE https://xxx/remote.php/dav/files/xxx/report-635.csv 403 (Forbidden)
(anonymous) @ xhr.js:195
xhr @ xhr.js:15
f @ dispatchRequest.js:51
Promise.then
_request @ Axios.js:163
request @ Axios.js:40
a.A.forEach.F.<computed> @ Axios.js:213
(anonymous) @ bind.js:5
z @ deleteUtils.ts:99
exec @ deleteAction.ts:58
oe @ actionUtils.ts:41
onActionClick @ FileEntryActions.vue:177
click @ FileEntryActions.vue:5
mn @ vue.runtime.esm.js:3033
n @ vue.runtime.esm.js:1832
mn @ vue.runtime.esm.js:3033
e.$emit @ vue.runtime.esm.js:3732
onClick @ actionText-fFcUPi2g.mjs:77
handleClick @ NcActionButton-CpGK6gSX.mjs:143
mn @ vue.runtime.esm.js:3033
n @ vue.runtime.esm.js:1832
Rr.i._wrapper @ vue.runtime.esm.js:7500
deleteAction.ts:62 [ERROR] files: Error while deleting a file {app: 'files', uid: 'dklein', level: 2, error: r, source: 'https://xxx/remote.php/dav/files/xxx/report-635.csv', …}
log @ index.mjs:54
error @ index.mjs:72
exec @ deleteAction.ts:62
await in exec
oe @ actionUtils.ts:41
onActionClick @ FileEntryActions.vue:177
click @ FileEntryActions.vue:5
mn @ vue.runtime.esm.js:3033
n @ vue.runtime.esm.js:1832
mn @ vue.runtime.esm.js:3033
e.$emit @ vue.runtime.esm.js:3732
onClick @ actionText-fFcUPi2g.mjs:77
handleClick @ NcActionButton-CpGK6gSX.mjs:143
mn @ vue.runtime.esm.js:3033
n @ vue.runtime.esm.js:1832
Rr.i._wrapper @ vue.runtime.esm.js:7500
```

</details>

--- a/lib/Operation.php
+++ b/lib/Operation.php
@@ -108,11 +108,18 @@ class Operation implements IComplexOperation, ISpecificOperation {
 		}
 
 		// '', admin, 'files', 'path/to/file.txt'
-		$segment = explode('/', $fullPath, 4);
+		// '', admin, 'files_trashbin', 'versions', 'path/to/file.txt'
+		$segment = explode('/', $fullPath, 5);
 
 		if (isset($segment[2]) && $segment[1] === '__groupfolders' && $segment[2] === 'trash') {
 			// Special case, a file was deleted inside a groupfolder
 			return true;
+		}
+
+		if (isset($segment[3]) && $segment[2] === 'files_trashbin'
+			&& ($segment[3] === 'keys' || !isset($segment[4]))) {
+			// Ignore encryption keys in trashbin and the files and version folder themselves
+			return false;
 		}
 
 		return isset($segment[2]) && in_array($segment[2], [
@@ -139,6 +146,25 @@ class Operation implements IComplexOperation, ISpecificOperation {
 				// Remove trailing ".v{timestamp}"
 				$innerPath = substr($innerPath, 0, -12);
 			}
+			return 'files/' . $innerPath;
+		} elseif ($folder === 'files_trashbin') {
+			// 'files', 'path/to/file.txt'
+			// 'versions', 'path/to/file.txt'
+			$segments = explode('/', $innerPath, 2);
+			if (isset($segments[1])) {
+				$innerPath = $segments[1];
+			}
+
+			if (preg_match('/.+\.d\d{10}$/', basename($innerPath))) {
+				// Remove trailing ".d{timestamp}" of trashbin
+				$innerPath = substr($innerPath, 0, -12);
+			}
+
+			if ($folder === 'versions' && preg_match('/.+\.v\d{10}$/', basename($innerPath))) {
+				// Remove trailing ".v{timestamp}" of versions inside trashbin
+				$innerPath = substr($innerPath, 0, -12);
+			}
+
 			return 'files/' . $innerPath;
 		} elseif ($folder === 'thumbnails') {
 			[$fileId,] = explode('/', $innerPath, 2);

--- a/tests/Integration/features/mimetypes.feature
+++ b/tests/Integration/features/mimetypes.feature
@@ -85,3 +85,16 @@
     And The webdav response should have a status code "403"
     When User "test1" copies file "/hello" to "/hello.txt"
     And The webdav response should have a status code "403"
+
+  Scenario: Blocking by mimetype works in trashbin
+    Given User "test1" uploads file "data/textfile.txt" to "/foobar.txt"
+    And user "admin" creates global flow with 200
+    | name      | Admin flow                       |
+    | class     | OCA\FilesAccessControl\Operation |
+    | entity    | OCA\WorkflowEngine\Entity\File   |
+    | events    | []                               |
+    | operation | deny                             |
+    | checks-0  | {"class":"OCA\\\\WorkflowEngine\\\\Check\\\\FileMimeType", "operator": "!is", "value": "httpd/unix-directory"} |
+    | checks-1  | {"class":"OCA\\\\WorkflowEngine\\\\Check\\\\FileMimeType", "operator": "!is", "value": "text/plain"} |
+    When User "test1" deletes file "/foobar.txt"
+    Then The webdav response should have a status code "204"

